### ISSUE TITLE
Benchmarks do not execute (as service process / VSTS agent) on non US systems

### ISSUE THREAD

**Issue by ChristophLindemann** [dotnet/BenchmarkDotNet#1487]
# What went wrong

When .NET Framework benchmarks are executed from a VSTS agent, running on a _**non US English**_ Windows installation, the benchmark fails with 

```
// **************************
// Benchmark: BlaBla
// *** Execute ***
// Launch: 1 / 1
// Execute: D:\agent\perf-win01\_work\3\s\src\blabla\bin\Release\net48\DefaultJob.exe --benchmarkName "BlaBla" --job "Default" --benchmarkId 0 in 
// BeforeAnythingElse
ExitCode != 0
// Benchmark Process 4344 has exited with code 255
No more Benchmark runs will be launched as NO measurements were obtained from the previous run!
```

```
System.NotSupportedException: Unknown Acknowledgment: ´╗┐Acknowledgment
   at BenchmarkDotNet.Engines.ConsoleHost.SendSignal(HostSignal hostSignal)
   at BenchmarkDotNet.Autogenerated.UniqueProgramName.AfterAssemblyLoadingAttached(String[] args)
   at BenchmarkDotNet.Autogenerated.UniqueProgramName.Main(String[] args)
```

_This problem does not occur for .NET Core, only for .NET Framework!_

Looking at a crashdump, it fails in [`public void SendSignal(HostSignal hostSignal)`](https://github.com/dotnet/BenchmarkDotNet/blob/0445917bf93059f17cb09e7d48cdb5e27a096c37/src/BenchmarkDotNet/Engines/ConsoleHost.cs#L25) at https://github.com/dotnet/BenchmarkDotNet/blob/0445917bf93059f17cb09e7d48cdb5e27a096c37/src/BenchmarkDotNet/Engines/ConsoleHost.cs#L33-L34

The `NotSupportedException` is thrown because it receives an `acknowledgment` string containing garbage at the start: **"´╗┐Acknowledgment" (bytes `B4 00 57 25 10 25 ...`)**

# Possible reason

I expect the reason to be a missmatch in the encoding / codepages used in the outer process _UTF-8 / 1200_ and in the inner job process _Western European (DOS) / 850_.

# Possible fix

In both [`SynchronousProcessOutputLoggerWithDiagnoser`](https://github.com/dotnet/BenchmarkDotNet/blob/2ba30330ec5ec8a9abe284c927277a6fe0fc5a0c/src/BenchmarkDotNet/Loggers/SynchronousProcessOutputLoggerWithDiagnoser.cs#L37-L62) and [`ConsoleHost`](https://github.com/dotnet/BenchmarkDotNet/blob/3223c94a92050147a02482a3810860703f0c5171/src/BenchmarkDotNet/Engines/ConsoleHost.cs#L13) wrap the StandartInput / StandardOutput with a TextReader/Writer that uses the _same_ encoding (UTF-8).

# Possibly related issues

- #1136 
- #1480 
- https://github.com/gfoidl/Base64/issues/139

**Comment by ChristophLindemann** [comment]
Also, please note that handling of the Encoding for the Console has changed from .NET Framework to .NET Core.

So this might also produce problems when then the outer process is .Net Core and the Job is .Net Framework (vice versa)

See
- https://github.com/dotnet/runtime/issues/17849
- https://github.com/dotnet/corefx/commit/42f9c2d4438ce9897c3d336ec52298afb4dc95fa

**Comment by adamsitnik** [comment]
Hi @ChristophLindemann 

thank you for a very detailed bug report.

I've sent a PR with a fix and integration tests: https://github.com/dotnet/BenchmarkDotNet/pull/1490

Our [CI has published](https://ci.appveyor.com/project/dotnetfoundation/benchmarkdotnet/builds/33799153/artifacts) an updated version of NuGet package. Could you please add our CI feed to your `nuget.config` file (can be created by running `dotnet new nugetconfig`) install `0.12.1.1382` and let me know if it works?

```xml
<packageSources>
  <add key="bdn-CI" value="https://ci.appveyor.com/nuget/benchmarkdotnet" />
</packageSources>
```

**Comment by lovettchris** [comment]
Perhaps this is a better fix?
https://github.com/dotnet/BenchmarkDotNet/pull/1491

**Comment by ChristophLindemann** [comment]
@adamsitnik 
I will give it a try tomorrow when I am back at my desk.

However I will be surprised if the ordinal comparison is enough, as I would expect the ordinal values of the “garbage” to be included in the comparison.

If it does not work, I think UTF8 would be quite safe, or even ASCII. Unless you have some very esoteric default code page setup. Or you could filter/trim any non ASCII codepoints.

Anyways, i’ll give it a try. But I can only test it for Scandinavian codepages.
I think someone should verify this on 2 byte systems like Japanese or Chinese.

**Comment by ChristophLindemann** [comment]
After looking a little more into the differences of .net core vs .net framework, I think the problem might actually be that: on Windows, the .net framework Encoding.Default maps to the active code page, while .net core always maps to UTF8 (Windows and Linux)

__But__, even if you set the active code page to UTF8 (using `chcp 65001`, as VSTS does) that codepage will not be used if the console is redirected, here it will fallback to `Encoding.Default` which is UTF8 _with BOM_.
So the garbage at the start is actually the UTF8 BOM.

So, @adamsitnik, I think the correct and safe thing, would be to force the encoding, of the redirected console, to UTF8 _without_ BOM. Something also suggested by @lovettchris

**Comment by ChristophLindemann** [comment]
@adamsitnik i have tried the 0.12.1.1382 package, and it indeed fixes the problem.

I still think, setting the encoding would be a more correct way to solve this, but I accept the current solution as it solves my problem.

When do you think this can be released?

**Comment by adamsitnik** [comment]
Fixed by #1491
